### PR TITLE
Disable 1D linking and Simultaneous Row Plotting in Mosviz

### PR DESCRIPTION
**Pull request by duytnguyendtn**
### Description

You may be asking yourself, why in {insert deity name of choice}'s name would you want that!? I think a picture is worth 1000 words (even if it's a picture of words :P)

image

You're reading that right; 200 targets, which used to take 1 hour and 11 minutes (!!) loaded in under two minutes.

This PR is part of a 3 part strategy in changing the linking strategy in Mosviz, developed with @camipacifici. In Mosviz, data navigation is mainly intended to be done with the Table Viewer. I recall us having discussions about removing the data dropdowns from Mosviz because of this. Though it may have been stopped due to a desire to plot multiple 1D spectrum in the same viewer.

After some profiling, I discovered supporting multi-plotting 1D spectrum was a VAST majority of the load time issues with Mosviz parser (in the GLASS dataset's case, 97.5% of the load time!). This is because the linking tree grows much faster than N for N number of targets, since we need to back link each target to each other one. This PR disables the intra 1D spectra linking to vastly improve loading times. It then hides other rows' data and disables the ability to load them simultaneously:

image

image

What about the multi-plotting spectrum? That's the last part of this 3-part strategy: (EDIT: Modified to include https://github.com/spacetelescope/jdaviz/pull/1790#issuecomment-1295130196)
1. Disable 1D linking
2. Investigate whether batch linking produces a similar benefit to removing linking entirely (https://github.com/spacetelescope/jdaviz/pull/1790#issuecomment-1295542053)
4. Re-enable the data dropdown with filtering to only the current row's data
5. Benchmark the remaining two minutes to see whether lazy data loading could help
6. If the above comes back true, investigate lazy data loading to further improve data loading/parsing time
7. Investigate Dynamic linking of data, and re-enable intra-row/simultaneous-row plotting

I propose we investigate the possibility of dynamic linking of data. That being we only link data together IF and WHEN the user actually requests to show the data at the same time. Effectively, when the user clicks on the data dropdown and tries to add a second spectrum to the spectrum viewer, link AT THAT TIME, not in advance. That way, we only end up with links that we actually need. Overall, this will result in a link tree that is "sub-optimal" as compared to one created with "batch loading". But what we sacrifice in optimal linking, we gain in efficiency and parsing speed. I have no idea if this is possible, as it's just an idea I came up with much past my bedtime. But here we are!

I got the green light from @camipacifici to proceed with 1 and 2 now, and investigate 3 later. So I invite eyes and comments!

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by kecnry** [comment]
I don't think disabling the data menu entirely is the way to go (won't be able to toggle the visibility of subsets or models, etc)... what if we instead skipped entries in the data menu from other rows?  Even then, that would disable the use-case of comparing spectra from different targets - do we want to do that?

**Comment by duytnguyendtn** [comment]
> Even then, that would disable the use-case of comparing spectra from different targets - do we want to do that?

Sorry for not being clear. This is what I meant by "multi-plotting spectra." This is exactly what Cami and I are proposing right now, since without 1D linking, this isn't possible. Cami and I agreed the "flipping through the table viewer" use case is probably more likely than the desire to compare spectra between targets, so we should remove the ability to do that from the UI, at least temporarily until we can figure out a way to allow it without crippling parse times. 

Good point on preserving the data dropdown for models and other "non data artifacts." I agree with your strategy of trying to remove entries from other rows; let me see if I can investigate

**Comment by duytnguyendtn** [comment]
Thanks team for the feedback this morning! Here's a summary of next steps:

1. Investigate whether batch linking produces a similar benefit to removing linking entirely
2. Re-enable the data dropdown with filtering to only the current row's data
3. Benchmark the remaining two minutes to see whether lazy data loading could help
4. If the above comes back true, investigate lazy data loading to further improve data loading/parsing time

**Comment by duytnguyendtn** [comment]
Ok so after a thorough archaeological expedition, I wanted to loop back on the first point:

> 1. Investigate whether batch linking produces a similar benefit to removing linking entirely

I swore I remembered at the time that this was already implemented, but I wanted to be sure before I embarrassed myself. So in https://github.com/spacetelescope/jdaviz/pull/762#discussion_r690784894, @ojustino does a good job describing the current strategy. Let me try to describe what I understand of the linking logic at this point:

The linking logic is already self contained within an entirely separate parser:
```
@data_parser_registry("mosviz-link-data")
def link_data_in_table(app, data_obj=None):
```
Linking is already deferred in each of the other parsers for the image, 1D, and 2D data. Only after those parsers have completed their work, does this `link parser` step in.

In this linking parser, we already employ `with app.data_collection.delay_link_manager_update():` to stop the automatic linking. We instead define the links manually by linking each 1D spectra to the first spectrum with a series of `LinkSame`'s. NOTICE: We aren't defining all the interlinks from each spectra to each other, only to the first spectrum. Also note that we aren't officially registering the link JUST YET. The logic appends all the `LinkSame`'s into a list containing all the pending links to be registered, and then ONLY AT THE END, applies all of them in a single `add_link()` call.  It's only when we add the links at the end to glue that the rest of the links are discovered and applied.

So this mosviz-link-parser I think is already our attempt to batch link. Searching through the glue docs, I can't find any other interpretation of "batch linking." Considering Tom's approval in https://github.com/spacetelescope/jdaviz/pull/762#discussion_r690877489, I think this produces the most efficient linking tree possible. I tried searching for some glue linking mechanism to provide an array of IDs and tell Glue "These are all the same thing," rather than creating separate `LinkSame`'s in a loop, but I couldn't find any. I think we're at the point to acknowledge we just have a lot of data to link and now need to reconsider which links we want to intentionally make, rather than blind linking everything

Please let me know if I'm misinterpreting "batch linking" and someone has a secret that I've been out of the loop on!

**Comment by pllim** [comment]
Looks like batch linking to me...

https://github.com/spacetelescope/jdaviz/blob/c319c30c7744f59c62aefc59f2860ee6eea94a79/jdaviz/configs/mosviz/plugins/parsers.py#L117-L120

File: jdaviz/components/viewer_data_select.vue
Comment (kecnry) [inline jdaviz/components/viewer_data_select.vue:201]: I'm not sure this will be sufficient... what happens if you create a model and unload it from the viewer (but its still assigned to the active row)?  Does it show in the "extra items" dropdown or does it disappear forever?

Comment (duytnguyendtn) [reply jdaviz/components/viewer_data_select.vue:201]: Aha! You've run into the second part of that commit! https://github.com/spacetelescope/jdaviz/pull/1790/commits/668a2bb1b65e441c5076ef4229dfe71f51ee01b0#diff-671bb80f9bf0db3b85cab9b0e24692dc85215aa08b106a50c148914e8d7a8e0fR145-R146

This is a second improvement I ran past @camipacifici; I'm basically hacking the functionality of the bug found fixed in #1808 / #1798 so that all data from the same row is always at the top (never in the dropdown). Therefore, your model, and all datasets assigned to the same row, will always be there, even though I removed the dropdown for the "other" datasets

Comment (duytnguyendtn) [reply jdaviz/components/viewer_data_select.vue:201]: Might be controversial, but figured I'd put it here and see what people think!

Comment (kecnry) [reply jdaviz/components/viewer_data_select.vue:201]: This now behaves differently than all the other configs, where items that are valid to load, but not currently loaded into the viewer, fall below the expanded menu.

image

I think we still want the expanded menu in this case (although the text that mentions "incl other rows" will need to be removed).

Comment (camipacifici) [reply jdaviz/components/viewer_data_select.vue:201]: Considering that the behavior common to the other configs is very annoying for Mosviz, I think there is no harm in having something different here. We can give it a try and revert or move to a new solution when a better approach is found.

Comment (duytnguyendtn) [reply jdaviz/components/viewer_data_select.vue:201]: Couple thoughts on the above:

1. Sorting would definitely be possible, though it would probably come at the cost of performance; we might need to loop through all the data entries every time before being able to sort, since i don't think we control the order the entries come in through the loop
2. We could also change it so that you can't "unload" the data of the same row, and can only toggle the visibility?

Comment (kecnry) [reply jdaviz/components/viewer_data_select.vue:201]: Sorry - I must be missing some context.  What is the annoying case we're trying to improve?  I just remember a lot of discussions with @Jenneh when designing the menu and if I remember correctly the idea of the expanded menu was to avoid user confusion about the difference between loading/unloading vs toggling visibility, and this entirely reverts that and only for mosviz.  If the icons/tooltips are sufficient to make that distinction, then I think we should remove the expanded menu across all tools.

Comment (duytnguyendtn) [reply jdaviz/components/viewer_data_select.vue:201]: The context here is if you generate data (say with gaussian smooth) and unload it, that spectrum gets dropped all the way to the bottom of the dropdown, requiring you to sift through all of the 1D spectra to find the data relevant to the current row.

https://user-images.githubusercontent.com/25206008/199738792-9e7523fb-5c99-4677-822e-18ce96bac134.mp4

This is the behavior Cami is describing as very annoying

Comment (kecnry) [reply jdaviz/components/viewer_data_select.vue:201]: Yes, but doesn't this PR remove all the other entries from other rows?  I agree this is an annoyance we should address if/when we re-introduce those rows, but would no longer be an issue with this PR in place, right?

Comment (duytnguyendtn) [reply jdaviz/components/viewer_data_select.vue:201]: The removal of the 1D links is intended to be a temporary solution. We intend to bring it back, hence why I'm thinking of this change in the context of the eventual goal of the 1D spectra returning. The solution we implement here should be compatible with the eventuality that the 1D spectra will come back.

(So a beneficial byproduct of the change here is if we don't put the row-relevant data in the dropdown, the solution here is as simple as the code you highlighted above. If not, then we'd have to add some additional filtering logic as you described.)

Comment (duytnguyendtn) [reply jdaviz/components/viewer_data_select.vue:201]: > If not, then we'd have to add some additional filtering logic

Not to say that this is even difficult. Though if the simpler solution also solves this annoyance, then I felt it was at least worth it to consider (hence why we're here 😄 )

Comment (kecnry) [reply jdaviz/components/viewer_data_select.vue:201]: Ok, can we attempt a different solution with sorting that isn't inconsistent with the other tools and see if there are actually performance issues (that is definitely a valid point)?

I could either spend a few hours playing with that outside of this PR (with all rows) to see if its feasible before we make the decision here, or we could defer until the rows are reintroduced (in which case I would personally prefer just removing the rows but keeping the expanding menu for now and keeping this solution in our back pocket as one option when we get to that rather than possibly switching back and forth).

Comment (camipacifici) [reply jdaviz/components/viewer_data_select.vue:201]: Sorry, I missed the information that rows are not to be reintroduced yet. I guess I will let you figure this out because my opinion is too much outside the code and we all agree what the end product should be. Thank you

Comment (duytnguyendtn) [reply jdaviz/components/viewer_data_select.vue:201]: Based on the team discussion, this change seems to require more thought. I'll change the filtering logic to preserve the "other" dropdown and we'll kick this can down the road. Thanks for the input, team!

**Comment by javerbukh** [comment]
I'm testing the NIRISS example notebook and getting the following error, anyone else? Traceback ends with:

```
File ~/Documents/jdaviz_dev/jdaviz/jdaviz/app.py:1031, in Application.add_data_to_viewer(self, viewer_reference, data_label, visible, clear_other_data)
   1013 """
   1014 Plots a data set from the data collection in the specific viewer.
   1015 
   (...)
   1027     defined data set.
   1028 """
   1029 if 'mosviz_row' in self.state.settings:
   1030     if not ((self.get_viewer("table-viewer").row_selection_in_progress) and
-> 1031             (self.data_collection[data_label].meta['mosviz_row'] !=
   1032                 self.state.settings['mosviz_row'])):
   1033         raise NotImplementedError("Intra-row plotting not supported. "
   1034                                   "Please use the table viewer to change rows")
   1036 viewer_item = self._get_viewer_item(viewer_reference)

KeyError: 'mosviz_row'
```

**Comment by duytnguyendtn** [comment]
@javerbukh that is an intentional, unrelated bug; see https://github.com/spacetelescope/jdaviz/issues/1816

**Comment by duytnguyendtn** [comment]
The nirspec parser properly assigns the rows, so the MosvizExample notebook shows the proper code path and proper error that this will show once we fix that bug

**Comment by duytnguyendtn** [comment]
Fixed row selection in https://github.com/spacetelescope/jdaviz/pull/1790/commits/4012700f3c3e6c640e7a2996f90eb8ea743f5cb9! You'll still see this traceback in NIRISS datasets when manually calling `add_data_to_viewer` with a different row's spectra due to #1805. But I think this can be handled as a separate bug

**Comment by javerbukh** [comment]
Works well! I did want to ask if one interaction behaves as expected: if I use gaussian smooth on a 1d spectrum, new data is added called "... smooth stddev x". If I make this data invisible or remove it from the viewer, then click a different row, then return to the first row, that "... smooth stddev x" data is now visible **and** in the viewer. Is this the behavior we would expect? If not, is it a big enough issue to handle in this PR or instead should we create a new issue for follow up? If it's a simple enough fix I think we should handle it in this PR but I'm interested to hear thoughts from the team and @camipacifici.

**Comment by duytnguyendtn** [comment]
@javerbukh Great eyes! That's yet another bug that Cami and I have been talking about, and plan to address. I was able to reproduce this error in `main` so it doesn't appear connected to the changes here

**Comment by duytnguyendtn** [comment]
That particular bug is https://github.com/spacetelescope/jdaviz/issues/1816. I don't know where this behavior is coming from, so it will require at least a little investigative work. I'd advocate for keeping it a separate effort

**Review by javerbukh — APPROVED** [review]
@duytnguyendtn Ah ok, in that case I think this is good to get in. Nice work!

File: CHANGES.rst
Comment (pllim) [inline CHANGES.rst:21]:
```suggestion
  to substantially decrease load times. [#1790]
```

File: jdaviz/app.py
Comment (pllim) [inline jdaviz/app.py:1032]: I think all 3 `if`s can be chained together with `and` into just 1 `if`?

File: jdaviz/configs/mosviz/tests/test_data_loading.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:242]:
```suggestion
```

File: jdaviz/app.py
Comment (duytnguyendtn) [reply jdaviz/app.py]: Duh... Thanks for catching that. I've clearly been staring at these lines for too long and have gotten loopy 😅 Done in https://github.com/spacetelescope/jdaviz/pull/1790/commits/f85fddebf1112f8951816f77797ebdc2de235d89

**Review by pllim — APPROVED** [review]
LGTM. Since @kecnry commented before, do you want an approval from him as well?

**Comment by duytnguyendtn** [comment]
@kecnry gave me a silent approval over slack, just couldn't do a full review. I'm happy to move forward here! Thanks Kyle, Jesse, Pey-Lian and Cami for your efforts/thoughts through this PR!

**Comment by kecnry** [comment]
> If I make this data invisible or remove it from the viewer, then click a different row, then return to the first row, that "... smooth stddev x" data is now visible and in the viewer.

> I don't immediately know where this behavior is coming from, so it will require at least a little investigative work.

The row changing removes/adds data to the viewer.  When data is removed from the viewer, all associated layers are destroyed/forgotten, so if we want any of the layer information to be restored when returning to the row, we will need to store that somewhere.  This same thing happens if you manually remove and add the same data entry into a viewer (all layer options are lost) which was the motivation for changing the check box in the data menu to control visibility instead of loaded-ness... but that solution obviously won't work for mosviz, so if the expected behavior is that visibility and/or layer state persists, we'll need to write that logic.

**Pull request merged**
